cmd/mps-control-daemon/main.go

```diff
@@ -0,0 +1,238 @@
+	if restarting {
+		err := stopDaemons(daemons...)
+		if err != nil {
+			return fmt.Errorf("error stopping plugins from previous run: %v", err)
+		}
+	}
+
+	klog.Info("Starting Daemons.")
+	daemons, restartDaemons, err := startDaemons(c, cfg, restarting)
+	if err != nil {
+		return fmt.Errorf("error starting plugins: %v", err)
+	}
+
+	if restartDaemons {
+		klog.Infof("Failed to start one or more plugins. Retrying in 30s...")
+		restartTimeout = time.After(30 * time.Second)
+	}
+
+	restarting = true
+	// Start an infinite loop, waiting for several indicators to either log
+	// some messages, trigger a restart of the plugins, or exit the program.
+	for {
+		select {
+		// If the restart timeout has expired, then restart the plugins
+		case <-restartTimeout:
+			goto restart
+
+		// Watch for any signals from the OS. On SIGHUP, restart this loop,
+		// restarting all of the plugins in the process. On all other
+		// signals, exit the loop and exit the program.
+		case s := <-sigs:
+			switch s {
+			case syscall.SIGHUP:
+				klog.Info("Received SIGHUP, restarting.")
+				goto restart
+			default:
+				klog.Infof("Received signal \"%v\", shutting down.", s)
+				goto exit
+			}
+		}
+	}
+exit:
+	if err := stopDaemons(daemons...); err != nil {
+		return fmt.Errorf("error stopping daemons: %v", err)
+	}
+	return nil
+}
+
+func startDaemons(c *cli.Context, cfg *Config, restarting bool) ([]*mps.Daemon, bool, error) {
+	// Load the configuration file
+	klog.Info("Loading configuration.")
+	config, err := cfg.loadConfig(c)
+	if err != nil {
+		return nil, false, fmt.Errorf("unable to load config: %v", err)
+	}
+	spec.DisableResourceNamingInConfig(logger.ToKlog, config)
+
+	// Update the configuration file with default resources.
+	klog.Info("Updating config with default resource matching patterns.")
+	err = rm.AddDefaultResourcesToConfig(config)
+	if err != nil {
+		return nil, false, fmt.Errorf("unable to add default resources to config: %v", err)
+	}
+
+	// Print the config to the output.
+	configJSON, err := json.MarshalIndent(config, "", "  ")
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to marshal config to JSON: %v", err)
+	}
+	klog.Infof("\nRunning with config:\n%v", string(configJSON))
+
+	// Get the set of daemons.
+	klog.Info("Retrieving daemons.")
+	daemonManager, err := mps.New(
+		mps.WithConfig(config),
+	)
+	if err != nil {
+		return nil, false, fmt.Errorf("error creating plugin manager: %v", err)
+	}
+	daemons, err := daemonManager.Daemons()
+	if err != nil {
+		return nil, false, fmt.Errorf("error getting daemons: %v", err)
+	}
+
+	// Loop through all daemons, starting them if they have any devices
+	// to serve. If even one plugin fails to start properly, try
+	// starting them all again.
+	started := 0
+	for _, p := range daemons {
+		// Just continue if there are no devices to serve for plugin p.
+		if len(p.Devices()) == 0 {
+			continue
+		}
+		// Start the MPS daemon
```

Comment:
Where is the check for MPS enabled in the config? I see we are using same nodeSelector as device-plugin (which is to run on every GPU node by default). How are we checking to make mps pod no-op when it is not enabled in teh config?

Reply:
Two things here:
* I have added a node selector that requires `nvidia.com/sharing.mps.enabled = true`
* We only instantiate a Daemon if sharing is enable: https://github.com/NVIDIA/k8s-device-plugin/pull/490/files#diff-061ca4ea886429b6d2e0db6954e0416cf178e0ddff0c0a426084b17c25feba2fR60 (This should be improved to properly detect MPS).

I will improve the readability though to ensure that it is clearer that there is a no-op if MPS is not actually enabled for a device.